### PR TITLE
Add ConsoleLink Creation/Remove

### PR DESCRIPTION
**Pull request by rashelrr**
Added ConsoleLink Creation/Remove
Added go mod vendor folder

### CODE REVIEW

File: pkg/helper/webconsole.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	err := c.Get(ctx, types.NamespacedName{Name: consoleLinkName}, consoleLink)
+	if err == nil && consoleLink != nil {
+		err = c.Delete(ctx, consoleLink)
+		if err != nil {
+			logger.Error(err, "Failed to delete the consolelink:", consoleLinkName)
+		} else {
+			logger.Info("Deleted the consolelink:", consoleLinkName)
+		}
+	}
+}
```
Comment (bmozaffa) [inline pkg/helper/webconsole.go:116]: Needs CR/LF

File: pkg/helper/webconsole.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	checkConsoleLink(ctx, consoleLinkName, apimanager, c)
+}
+
+func checkConsoleLink(ctx context.Context, consoleLinkName string, apimanager *appsv1alpha1.APIManager, c client.Client) {
+	route := getRoute(ctx, c, apimanager)
+	consoleLink := &consolev1.ConsoleLink{}
+	err := c.Get(ctx, types.NamespacedName{Name: consoleLinkName}, consoleLink)
+	if route != nil {
+		if err != nil && apierrors.IsNotFound(err) {
+			consoleLink = createNamespaceDashboardLink(consoleLinkName, route, apimanager)
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/helper/webconsole.go:43]: Which function name are you referring to?

[the remark this answers is not in the document]
Comment (bmozaffa) [reply pkg/helper/webconsole.go:43]: createNamespaceDashboardLink, in that it only creates an object in memory. Not a big deal, but maybe try to make that clear in the name.

[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/helper/webconsole.go:43]: I'm basing my code off of other consolelink creation examples (Teiid, Apicurio) to be consistent, but I can change the name no problem

[the remark this answers is not in the document]
Comment (bmozaffa) [reply pkg/helper/webconsole.go:43]: This one is not a big deal, so you could go either way, but as far as consistency, it only matters that the code is consistent with other parts of this operator.

[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/helper/webconsole.go:43]: Got it okay

File: deploy/cluster_role.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  name: 3scale-operator
+rules:
+  - apiGroups:
+      - console.openshift.io
+    resources:
+      - consolelinks
+    verbs:
+      - create
+      - delete
+      - get
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply deploy/cluster_role.yaml]: Done!

File: pkg/controller/apimanager/apimanager_controller.go [hunk 4]
```diff
@@ -3,29 +3,31 @@ package apimanager
 import (
 	"context"
 	"fmt"
-	"reflect"
-	"sort"
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/controller/apimanager/apimanager_controller.go]: Done!

File: pkg/controller/apimanager/apimanager_controller.go [hunk 5]
```diff
@@ -139,8 +146,8 @@ func (r *ReconcileAPIManager) upgradeAPIManager(cr *appsv1alpha1.APIManager) (re
 func (r *ReconcileAPIManager) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	logger := r.Logger().WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	logger.Info("ReconcileAPIManager", "Operator version", version.Version, "3scale release", product.ThreescaleRelease)
-
 	instance, err := r.apiManagerInstance(request.NamespacedName)
+
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/controller/apimanager/apimanager_controller.go]: Fixed!

File: pkg/controller/apimanager/apimanager_controller.go [hunk 6]
```diff
@@ -139,8 +146,8 @@ func (r *ReconcileAPIManager) upgradeAPIManager(cr *appsv1alpha1.APIManager) (re
 func (r *ReconcileAPIManager) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	logger := r.Logger().WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	logger.Info("ReconcileAPIManager", "Operator version", version.Version, "3scale release", product.ThreescaleRelease)
-
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/controller/apimanager/apimanager_controller.go]: Fixed!

File: pkg/controller/apimanager/apimanager_controller.go [hunk 7]
```diff
@@ -204,17 +211,27 @@ func (r *ReconcileAPIManager) Reconcile(request reconcile.Request) (reconcile.Re
 func (r *ReconcileAPIManager) apiManagerInstance(namespacedName types.NamespacedName) (*appsv1alpha1.APIManager, error) {
 	// Fetch the APIManager instance
 	instance := &appsv1alpha1.APIManager{}
-
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/controller/apimanager/apimanager_controller.go]: Fixed!

File: pkg/controller/apimanager/apimanager_controller.go [hunk 8]
```diff
@@ -204,17 +211,27 @@ func (r *ReconcileAPIManager) Reconcile(request reconcile.Request) (reconcile.Re
...
 	err := r.Client().Get(context.TODO(), namespacedName, instance)
+
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/controller/apimanager/apimanager_controller.go]: Fixed!

File: pkg/helper/webconsole.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	"strings"
+)
+
+var logger = logs.GetLogger("openshift-webconsole")
+
+// ConsoleLinkText is the text of the consoleLink shown on the webconsole
+const ConsoleLinkText = "APIManager - 3scale"
+
+// ConsoleLinkExists checks if a ConsoleLink resource exists
+func ConsoleLinkExists() error {
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/helper/webconsole.go]: Fixed!

File: pkg/helper/webconsole.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+)
+
+var logger = logs.GetLogger("openshift-webconsole")
+
+// ConsoleLinkText is the text of the consoleLink shown on the webconsole
+const ConsoleLinkText = "APIManager - 3scale"
+
+// ConsoleLinkExists checks if a ConsoleLink resource exists
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/helper/webconsole.go]: Fixed!

File: pkg/helper/webconsole.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+func ConsoleLinkExists() error {
+	gvk := schema.GroupVersionKind{Group: "console.openshift.io", Version: "v1", Kind: "ConsoleLink"}
+	return kubernetes.CustomResourceDefinitionExists(gvk)
+}
+
+// CreateConsoleLink creates a ConsoleLink object if it doesn't already exist
+func CreateConsoleLink(ctx context.Context, c client.Client, apimanager *appsv1alpha1.APIManager) {
+	consoleLinkName := fmt.Sprintf("%s-%s", apimanager.ObjectMeta.Name, apimanager.Namespace)
+	checkConsoleLink(ctx, consoleLinkName, apimanager, c)
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/helper/webconsole.go]: Fixed!

File: pkg/helper/webconsole.go [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+}
+
+func checkConsoleLink(ctx context.Context, consoleLinkName string, apimanager *appsv1alpha1.APIManager, c client.Client) {
+	route := getRoute(ctx, c, apimanager)
+	consoleLink := &consolev1.ConsoleLink{}
+	err := c.Get(ctx, types.NamespacedName{Name: consoleLinkName}, consoleLink)
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/helper/webconsole.go]: Fixed!

File: pkg/helper/webconsole.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/helper/webconsole.go]: Fixed! Function name is now createConsoleLinkPointer.

File: pkg/apis/addtoscheme_apps_v1alpha1.go [hunk 13]
```diff
@@ -2,9 +2,15 @@ package apis
...
 
 func init() {
 	// Register the types with the Scheme so the components can map objects to GroupVersionKinds and back
-	AddToSchemes = append(AddToSchemes, v1alpha1.SchemeBuilder.AddToScheme)
-}
+	AddToSchemes = append(AddToSchemes,
+		routev1.Install,
+		consolev1.Install,
+		v1alpha1.SchemeBuilder.AddToScheme)
+}
```
[the remark this answers is not in the document]
Comment (rashelrr) [reply pkg/apis/addtoscheme_apps_v1alpha1.go]: Fixed!

File: pkg/helper/webconsole.go (hunk 1 above)
Comment (rashelrr) [reply pkg/helper/webconsole.go]: Fixed!

**Comment by miguelsorianod** [comment]
Hi,

Thanks for your contribution.

We are currently evaluating/analysing the functionality and the pull request and we will come back to you as soon as possible with some feedback

**Comment by eguzki** [comment]
@rashelrr before going into deep review, two quick comments:
* Do not add vendoring to the repo. `go.mod` file provides reproducible builds and adding vendor files is not necessary at all. 
* Do not update `go.mod` file unless it is necessary. I do not see any new dep added or removed. Did you need to "upgrade" existing dep? If so, could you explain briefly which one and why?

**Comment by eguzki** [comment]
Sure @bmozaffa 
Thnx for the info.

**Comment by myeung18** [comment]
> @eguzki Rashel was our summer intern who completed her internship last week. @myeung18 worked with her on this and other PRs, and would be able to pick this up and drive it home. However, he is on PTO this week and returns next week. Can we leave this PR alone and neither merge or close it, and let Marco open a new PR on top of these commits? That way, he can address these and other feedbacks in a new commit, and when resolved and merged, this PR will also automatically be marked as merged, so we can get Rashel's commits in. Does that work?


@eguzki   @bmozaffa 
A new PR#462 is created to replace this one

**Pull request closed without merging**